dartagnan/src/main/java/com/dat3m/dartagnan/program/event/core/InstructionBoundary.java

```diff
@@ -0,0 +1,54 @@
+package com.dat3m.dartagnan.program.event.core;
+
+import com.dat3m.dartagnan.program.event.AbstractEvent;
+import com.dat3m.dartagnan.program.event.Event;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.google.common.base.Preconditions.checkState;
+
+/**
+ * Instances of this class occur in pairs, where the link is owned by the later event.
+ * The "begin" boundary must be a dominator of the "end" boundary, but the opposite may not be true.
+ *
+ * <p>
+ * Visible events enclosed by such a pair are said to originate from the same instruction.
+ * In execution graphs, they shall be {@code si}-related.
+ */
+public final class InstructionBoundary extends AbstractEvent {
+
+    private final InstructionBoundary begin;
+
+    public InstructionBoundary(Void ignore, InstructionBoundary b) {
```

Comment:
Why do we need this first parameter?

Reply:
To avoid collision with the copy constructor.